escrow/src/main.sw

```diff
@@ -1,257 +1,355 @@
+    #[storage(read, write)]fn propose_arbiter(arbiter: Arbiter, identifier: u64) {
+        // The assertions ensure that only the seller can propose a new arbiter and the arbiter
+        // cannot be the buyer / seller, the arbiter will be able to take a none-zero payment
 
-                storage.seller.deposited = false;
-                storage.seller.approved = false;
+        let escrow = storage.escrows.get(identifier);
 
-                transfer_to_output(storage.asset_amount, storage.asset, storage.seller.address);
-            }
-        } else {
-            revert(0);
-        };
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+
+        let user = msg_sender().unwrap();
 
-        true
+        require(user == escrow.seller.address, UserError::Unauthorized);
+        require(arbiter.address != escrow.buyer.address, ArbiterInputError::CannotBeBuyer);
+        require(arbiter.address != escrow.seller.address, ArbiterInputError::CannotBeSeller);
+        require(0 < arbiter.fee_amount, ArbiterInputError::FeeCannotBeZero);
+        require(arbiter.fee_amount == msg_amount(), ArbiterInputError::FeeDoesNotMatchAmountSent);
+        require(arbiter.asset == msg_asset_id(), ArbiterInputError::AssetDoesNotMatch);
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
+        let proposal = storage.arbiter_proposal.get(identifier);
+        if proposal.is_some() && 0 < proposal.unwrap().fee_amount {
+            transfer(proposal.unwrap().fee_amount, proposal.unwrap().asset, escrow.seller.address);
+        }
+
+        storage.arbiter_proposal.insert(identifier, Option::Some(arbiter));
+
+        log(ProposedArbiterEvent {
+            arbiter, identifier
+        });
     }
 
-    /// Returns the amount of the specified asset in this contract
-    #[storage(read)]fn get_balance() -> u64 {
-        this_balance(storage.asset)
+    #[storage(read, write)]fn resolve_dispute(identifier: u64, payment_amount: u64, user: Identity) {
+        // The assertions ensure that a resolution can only occur during a dispute and only once
+        // by the specified arbiter. The deposit will be sent to either the buyer or seller and the
+        // arbiter can choose their payment amount up to the deposit from the seller
+
+        let mut escrow = storage.escrows.get(identifier);
+
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(escrow.disputed, StateError::NotDisputed);
+        require(msg_sender().unwrap() == escrow.arbiter.address, UserError::Unauthorized);
+        require(user == escrow.buyer.address || user == escrow.seller.address, UserInputError::InvalidRecipient);
+        require(escrow.buyer.asset.is_some() && 0 < escrow.buyer.deposited_amount, StateError::CannotResolveBeforeDesposit);
+        require(payment_amount <= escrow.arbiter.fee_amount, ArbiterInputError::PaymentTooLarge);
```

Comment:
IMO we should always just pay the arbiter their max fee amount, and then we don't have to worry about sending the seller the difference.

Reply:
That's a subjective choice for a feature. I have opted to allow the arbiter to pick their fee instead of limiting it to more basic functionality which transfers the full amount in all cases.